// components/FileBrowser/Directory.js
import React from 'react';
import PropTypes from 'prop-types';
import { Label, List } from 'semantic-ui-react';
import { If, Then, Else } from 'react-if';
import File from './File';
import yeast from 'yeast';

class Directory extends React.Component {
	static propTypes = {
		uuid: PropTypes.string,
		name: PropTypes.string,
		content: PropTypes.array,
	};

	static defaultProps = {
		uuid: yeast(),
		name: '.',
		content: [],
	};
	constructor(props) {
		super(props);
		this.state = { isOpen: props.isOpen || false };
	}
	handleOnClick = () => {
		this.setState({
			isOpen: !this.state.isOpen,
		});
	};
	render() {
		const { uuid, content, name } = this.props;
		const { isOpen } = this.state;
		const body = content.map(
			p => (p.isDirectory ? <Directory key={p.name + uuid} {...p} /> : <File key={p.name + uuid} {...p} />)
		);
		return (
			<List.Item>
				<List.Icon name={isOpen ? 'folder open outline' : 'folder outline'} onClick={this.handleOnClick} />
				<List.Content>
					<List.Header onClick={this.handleOnClick}>
						{name}
					</List.Header>
					<If condition={isOpen}>
						<Then>
							<List.List>
								{body}
							</List.List>
						</Then>
					</If>
				</List.Content>
			</List.Item>
		);
	}
}

export default Directory;
