// components/ZippedFileBrowser/index.js
import React from 'react';
import PropTypes from 'prop-types';
import FileBrowser from '../FileBrowser';
import JSZip from 'jszip';
import yeast from 'yeast';
import {
	head,
	merge,
	isObject,
	mergeWith,
	isArray,
	initial,
	size,
	toString,
	reduce,
	split,
	keys,
	trimEnd,
	zipWith,
	flatten,
	flatMap,
	set,
	get,
} from 'lodash';

const ROOT = '(ROOT)';

export default class ZippedFileBrowser extends React.Component {
	static propTypes = {
		data: PropTypes.string.isRequired,
	};
	state = {};
	// see https://stackoverflow.com/questions/16245767/creating-a-blob-from-a-base64-string-in-javascript
	b64toBlob(b64Data, contentType = '', sliceSize = 512) {
		var byteCharacters = atob(b64Data);
		var byteArrays = [];

		for (var offset = 0; offset < byteCharacters.length; offset += sliceSize) {
			const slice = byteCharacters.slice(offset, offset + sliceSize);

			let byteNumbers = new Array(slice.length);
			for (var i = 0; i < slice.length; i++) {
				byteNumbers[i] = slice.charCodeAt(i);
			}

			const byteArray = new Uint8Array(byteNumbers);

			byteArrays.push(byteArray);
		}

		return new Blob(byteArrays, { type: contentType });
	}
	baseName(str) {
		var base = new String(str).substring(str.lastIndexOf('/') + 1);
		if (base.lastIndexOf('.') != -1) base = base.substring(0, base.lastIndexOf('.'));
		return base;
	}
	splitSeperator(str) {
		return split(str, /[\\/]/);
	}
	trimSeperator(str) {
		return trimEnd(str, '/\\');
	}
	toFileBrowserEntries(zipFiles) {
		let res = {};
		let root = '';
		keys(zipFiles).map(key => {
			console.log({ res });
			const val = zipFiles[key];
			const name = this.trimSeperator(`${ROOT}/` + key);
			const splt = this.splitSeperator(name);
			const path = initial(splt);
			let elem;
			if (val.dir) {
				const baseName = this.baseName(name);
				elem = {
					uuid: yeast(),
					path: key,
					name: baseName,
					isDirectory: true,
					isOpen: true,
					date: val.date,
					content: [],
				};
			} else {
				const fileName = this.baseName(name);
				if (fileName === '.DS_Store') {
					return;
				}
				const dirName = this.trimSeperator(name.substr(0, name.lastIndexOf(fileName)));
				elem = {
					uuid: yeast(),
					path: key,
					isDirectory: false,
					name: fileName,
					date: val.date,
				};
			}
			// not correct
			const selector = flatMap(path, p => [p, 'content']);
			let e = get(res, selector, []);
			if (!isArray) {
				e = [elem];
			} else {
				e.push(elem);
			}
			console.log({ key, selector, e, res });
			set(res, selector, e);
		});
		res = res[`${ROOT}`].content;
		console.log({ res });
		return res;
	}

	async componentDidMount() {
		const zip = new JSZip();
		const blob = this.b64toBlob(this.props.data, 'text/plain');
		let zipHandle;
		try {
			zipHandle = await zip.loadAsync(blob);
		} catch (err) {
			this.setState({
				error: err.toString(),
			});
			return;
		}
		this.setState({
			files: this.toFileBrowserEntries(zipHandle.files),
		});
	}
	render() {
		const { files, err } = this.state;
		if (err) {
			return <div>Error</div>;
		}
		if (size(files) > 0) {
			return <FileBrowser content={files} />;
		}
		return <div>XXXX</div>;
	}
}
